### PR TITLE
Migrate current Heartbeat to new RPC module

### PR DESCRIPTION
**Pull request by gczsjdy**
## What changes were proposed in this pull request?

This is to migrate the current Heartbeat's sending Cache/Index/Metrics implementation(by changing Spark code) to new RPC module.

## How was this patch tested?

`FiberSensorSuite`

### CODE REVIEW

**Comment by gczsjdy** [comment]
@carsonwang The new commit can solve #678

**Comment by carsonwang** [comment]
Can you please make a separated PR to resolve #678?

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberSensorSuite.scala [hunk 1]
```diff
@@ -89,6 +88,8 @@ class FiberSensorSuite extends QueryTest with SharedOapContext
       }
     }
 
+    FiberCacheManager.registerHeartbeat()
```
Comment (carsonwang) [inline src/test/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberSensorSuite.scala]: Should this be registered already? Can we avoid this?

File: src/main/scala/org/apache/spark/SparkEnv.scala [hunk 2]
```diff
@@ -379,7 +379,7 @@ object SparkEnv extends Logging {
     val oapRpcDriverEndpoint = registerOrLookupEndpoint(
       OapRpcManagerMaster.DRIVER_ENDPOINT_NAME, oapRpcManagerMasterEndpoint)
 
-    val oapRpcManager = if (isDriver) {
+    val oapRpcManager = if (isDriver && !isLocal) {
```
[the remark this answers is not in the document]
Comment (carsonwang) [reply src/main/scala/org/apache/spark/SparkEnv.scala]: If it is local, only a slave will be created and no heartbeat will occur? It seems the master should always be created.

File: src/test/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberSensorSuite.scala (hunk 1 above)
Comment (carsonwang) [reply src/test/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberSensorSuite.scala]: any clue on this? This is not expected to be registered here.

**Comment by linhong-intel** [comment]
@gczsjdy @carsonwang 
any update on this PR?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala [hunk 3]
```diff
@@ -200,6 +196,18 @@ private[oap] class OapDataReader(
 
   import org.apache.spark.sql.execution.datasources.oap.INDEX_STAT._
 
+  registerHeartbeat()
```
Comment (carsonwang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataReaderWriter.scala]: This is not the right place to registerHeartbeat - Each reader is going to register.

**Comment by carsonwang** [comment]
Merging this. The only concern is the place to start heart beat. Ideally it should be started inside the OapRpcManagerSlave. Will have followup to see if it can be addressed.

**Comment by gczsjdy** [comment]
It's indeed ideal to start it automatically, while the problem is that some state are not ready at that time.

Probably we can make this better if we introduce something like `OapContext`, we know that all Spark related class(like `SparkContext`, etc) has been ready.

**Pull request merged**